### PR TITLE
Introduce requirement setup for integrations

### PR DESCRIPTION
**Pull request by cosenal**
## Description

At the moment, we don't pin the version of integration packages (Braket, Qiskit, etc.)  in the Mitiq package that we ship to users. This causes incompatibility issues, e.g, https://github.com/unitaryfund/mitiq/issues/2208.

This PR introduces integration packages as extra dependencies.

Example: if a user wants to use Mitiq with Qiskit, they can run 
```
pip install mitiq[qiskit]
```
in their enviroment, and similarly for all other integrations we support. This will guarantee that the version of the integration packages we support are compatible with the ones in the user's environment. 

🎪 One important bonus is that now the main Mitiq package only depends on `cirq-core` as opposed to the full `cirq` package 🎪

-----

This fixes https://github.com/unitaryfund/mitiq/issues/1201 (see Issue description)

### Docs
Readme has been updated.

Note: There will be a lapse of time between merging this PR and the next release, when the README will not be synced with reality of the package. We may consider split the Readme change and ship that along with the release PR. (cc @FarLab as milestone manager.) 

### Testing
I tested that:
1. for Mitiq developers, everything works just like before with `pip install -e .[development]`
2. the code snippet from the README works with `pip install mitiq` (without any extra)
3. the different extra installations for the supported frontend also works (tested started from empty Conda environments)

### CODE REVIEW

**Comment by jordandsullivan** [comment]
Good idea. I've already hit this issue several times.

**Review by natestemen — COMMENTED** [review]
On a high level this looks great. I like this structure a lot! 👍🏻

**Comment by cosenal** [comment]
@natestemen @jordandsullivan This is now ready to review. Please take a look.

File: setup.py [hunk 1]
```diff
@@ -22,13 +25,21 @@
 with open("mitiq/_version.py", "w") as f:
     f.write(f"__version__ = '{__version__}'\n")
 
+third_party_integration_requirements = {
```
Comment (jordandsullivan) [inline setup.py:28]: To confirm, when a user runs `pip install -e .[development]` or `make install`, this will install all the third party integrations as well, correct?

Comment (jordandsullivan) [reply setup.py:28]: How are we enforcing the version compatibility of the various 3P requirements? Do pip or setuptools do this automatically?

File: README.md [hunk 2]
```diff
@@ -96,12 +96,23 @@ We refer to any programming language you can write quantum circuits in as a _fro
...
 
-Note: Cirq is a core requirement of Mitiq and is installed when you `pip install mitiq`.
+You can install Mitiq support for these frontends by specifying them during installation, 
+as optional extras, along with the main package.
+To install Mitiq with one or more frontends, you can specify each frontend in square brackets as part of the installation command. 
+
+For example, 
+to install Mitiq with support for Qiskit and Qibo:
+```bash
+pip install mitiq[qiskit,qibo]
```
Comment (jordandsullivan) [inline README.md:110]: Is there a way to test this functionality before deploying this version on PyPI?

File: requirements/requirements-braket.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+amazon-braket-sdk~=1.69.0
+cirq-ionq>=1.0.0,<1.4.0
```
Comment (jordandsullivan) [inline requirements/requirements-braket.txt:2]: Why is `cirq-ionq` required for the `braket` integration?

File: requirements/requirements-cirq.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# This file is redundant, since the internal representation of choice is Cirq.
```
Comment (jordandsullivan) [inline requirements/requirements-cirq.txt:1]: I would rephrase just to make this a little clearer. Something like, "...since Cirq is the core internal representation for quantum circuits in Mitiq"

**Review by jordandsullivan — COMMENTED** [review]
If we have updated `make install` to include all these requirements, can we also update [this line](https://github.com/unitaryfund/mitiq/blob/setup-integrations/CONTRIBUTING.md?plain=1#L25) in the main CONTRIBUTING guide to say `make install` rather than `pip install -e ".[development]"`?

File: requirements/requirements-qibo.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+qibo==0.2.7 # TODO: unpin this
```
[the remark this answers is not in the document]
Comment (cosenal) [reply requirements/requirements-qibo.txt:1]: Notice I have just moved those dependencies around, namely I split them in multiple files, I haven't changed them.
The "unpin this" comes from https://github.com/unitaryfund/mitiq/pull/2221

File: requirements/requirements-braket.txt (hunk 3 above)
Comment (cosenal) [reply requirements/requirements-braket.txt:2]: Because of this: https://github.com/unitaryfund/mitiq/blob/8c24a9bddd4d6213cdd39fae2d750e55247e82a9/mitiq/interface/mitiq_braket/conversions.py#L8

File: requirements/requirements-cirq.txt (hunk 4 above)
Comment (cosenal) [reply requirements/requirements-cirq.txt:1]: Yes, better

File: setup.py (hunk 1 above)
Comment (cosenal) [reply setup.py:28]: > this will install all the third party integrations as well, correct?
Yes, exactly the same as before.

> How are we enforcing the version compatibility of the various 3P requirements?
Same as we do now, that is, having all of them together in the development environment and running the test suite in such environment.

File: README.md (hunk 2 above)
Comment (cosenal) [reply README.md:110]: Yes. To replicate the same example, you can run `pip install -e '.[qiskit,qibo]'` locally.

**Comment by cosenal** [comment]
@jordandsullivan Thanks for the thorough review. I addressed / replied to all your comments. PTAL

**Review by jordandsullivan — APPROVED** [review]
Thanks, looks good to me!

**Review by natestemen — APPROVED** [review]
Thanks Alessandro!

Looking forward to this helping users use Mitiq with less dependency issues.

I was brainstorming potential issues that could still arise, and I came up with the following scenario.

1. User gets set up in their desired frontend (not cirq), writing ciruits and running them
2. They learn about Mitiq
3. They run `pip install mitiq` without looking through the docs too much
4. They hit compatibility issues

If this users then come to us, it should be easy to help solve their problem, as we can ask them to run `pip install mitiq[frontend]`.

This is not a blocker, but I just wanted to write this down both for my understanding[^1], and so we all know there are still _some_ potential issues this does not solve. I don't think there's any magic bullet for this problem, however.

[^1]: If I'm misunderstanding something, let me know!

**Comment by cosenal** [comment]
@natestemen You're right, that's a valid scenario. One way to mitigate it would be to use the tools from `pkg_resources` to check at runtime (as early as possible) for the dependencies installed in the user environment. Example:
```python3
pkg_resources.working_set.require('qiskit~=2.0.0')

VersionConflict: (qiskit 1.0.2 (/opt/miniconda3/envs/mitiqenv/lib/python3.11/site-packages), Requirement.parse('qiskit~=2.0.2')) 
```
vs.
```python3
pkg_resources.working_set.require('qiskit~=1.0.2')

[qiskit 1.0.2 (/opt/miniconda3/envs/mitiqenv/lib/python3.11/site-packages)
...
]
```

Perhaps something for a follow-up PR, subject to prioritization, in case we get more dependency issues reported.

**Comment by natestemen** [comment]
Yup agreed, lets only pursue further ideas if we get more complaints. Thanks for jotting down a potential next step :)

**Comment by cosenal** [comment]
Merged this. After the two reviewers approvals, I have only merged main into the branch and solved the merge conflicts.
